### ISSUE TITLE
Duplicate tasks being added to the queue

### ISSUE THREAD

**Issue by matthieunelmes** [doowb/firebase-cron#10]
I seem to be having an issue where a task is added, sometimes 6x to the firebase-queue.

I really cannot understand how or why. This has only just started happening after running flawlessly for over a year.

I have hourly jobs running with the pattern "0 * * * *" However, the tasks are added 4 sometimes 6 times.

To try debugging, I set up a local test environment running 4 jobs every 60 seconds. However, using the cron.run() and displayJobs as per the docs I'm seeing the following in the console suggesting that the jobs are added twice every minute?

Any ideas?

[ Fri Oct 12 2018 21:36:00 GMT+0100 (BST) ]
-LNkB0ZS9iXTz0mh7djt 2018-10-12T20:36:00.000Z
-LNkB0ZYCktBCsrCOmXm 2018-10-12T20:36:00.000Z
-LNkB0Zfj7eV4_TJOxZd 2018-10-12T20:36:00.000Z
-LNkB0Zmeg29MmZOiH_v 2018-10-12T20:36:00.000Z
------------------------------

[ Fri Oct 12 2018 21:36:00 GMT+0100 (BST) ]
-LNkB0ZS9iXTz0mh7djt 2018-10-12T20:36:00.000Z
-LNkB0ZYCktBCsrCOmXm 2018-10-12T20:36:00.000Z
-LNkB0Zfj7eV4_TJOxZd 2018-10-12T20:36:00.000Z
-LNkB0Zmeg29MmZOiH_v 2018-10-12T20:36:00.000Z
------------------------------

[ Fri Oct 12 2018 21:37:00 GMT+0100 (BST) ]
-LNkB0ZS9iXTz0mh7djt 2018-10-12T20:37:00.000Z
-LNkB0ZYCktBCsrCOmXm 2018-10-12T20:37:00.000Z
-LNkB0Zfj7eV4_TJOxZd 2018-10-12T20:37:00.000Z
-LNkB0Zmeg29MmZOiH_v 2018-10-12T20:37:00.000Z
------------------------------

[ Fri Oct 12 2018 21:37:00 GMT+0100 (BST) ]
-LNkB0ZS9iXTz0mh7djt 2018-10-12T20:37:00.000Z
-LNkB0ZYCktBCsrCOmXm 2018-10-12T20:37:00.000Z
-LNkB0Zfj7eV4_TJOxZd 2018-10-12T20:37:00.000Z
-LNkB0Zmeg29MmZOiH_v 2018-10-12T20:37:00.000Z
------------------------------

**Comment by doowb** [comment]
Can you show the code for adding the job and when you call `run` and `displayJobs`? Also, which version of `firebase-cron` are you using?

Thanks.

**Comment by matthieunelmes** [comment]
Just checked and I'm on firebase-cron v1.0.1

To add cron jobs I'm using:

    const cron = new Cron(firebase.database().ref(), firebase.database().ref('queue'), {
        interval: 10000,
        endpoint: 'cron-jobs'
      });
      cron.addJob(imageId, '0 * * * *', {
        imageId: imageId,
        type: 'imageSnapShot',
        cronTaskId: imageId,
        hourlyCount: 0,
        dailyCount: 0
      }, function (err) {
        if (err) {
          //Error handling here
        }
      });

Then to monitor them and add tasks to the queue, on a separate server I'm using:

      const cron = new Cron(firebase.database().ref(), firebase.database().ref('queue'), {
        interval: 20000,
        endpoint: 'cron-jobs'
      });
      cron.run(function (jobs) {
        var now = cron.remoteDate();
          if (!jobs) {
            // console.log('[ ' + new Date(now) + ' ]' + ' no jobs');
            return;
          }
        console.log('[ ' + new Date(now) + ' ]');
        Object.keys(jobs).forEach(function (name) {
          var job = jobs[name];
          console.log(name, new Date(job.nextRun));
        });
        console.log('------------------------------');
        console.log();
    }, function (err) {
      console.log('Error running cron jobs : ', err);
    });
      console.log('Cron Watcher Started');

I'm fairly certain I'm not....but could it be that there are multiple server instances running the cron.run()? This was the only scenario I could think of why mulitple tasks would get added to the queue?

**Comment by doowb** [comment]
Thanks for the additional information

> firebase-cron v1.0.1

In v1, the api changed to return promises (so the `addJob` method would return a promise instead of taking a callback). Since you're running the `addJob` on a different server, then this shouldn't be an issue anyway.

The `run` call also looks fine. What's happening is that the callback is being executed twice with the same data. To determine if the jobs as being added to the firebase queue, you'd need to look under `queue/tasks` in your firebase account to see if there are 8 tasks being added each minute.

I've had a queue running for years and haven't had an issue like this, but I don't know if I started using v1 yet. I'll see if I can reproduce this when I have a chance.

**Comment by matthieunelmes** [comment]
Thanks for taking the time to reply.

It's a fantastic tool and as I say, I've also been running it for over a year without issue.

I did migrate the node app to a new Heroku instance, and in the process, it may have grabbed the most recent v1.0.1 package during node module install.

I'll run a few more tests today and report back.

Many thanks!

**Comment by matthieunelmes** [comment]
Ok, I'm getting closer to getting to the root of the problem. However I'm still slightly baffled.

It appears to be the way my Queue Worker handles the tasks once they've been placed on the queue.

So far example:

    var queue = new Queue(ref, function(data, progress, resolve, reject) {
	setTimeout(function() {
		progress(50);
	}, 1000);
	setTimeout(function() {
		resolve();
	}, 1000);
    });

Above works as expected. However whenever I reference back to the data within the cron-task to do something such as incrementing a count for how many times the task has run, for example:

    var queue = new Queue(ref, function(data, progress, resolve, reject) {
	setTimeout(function() {
		progress(50);
	}, 1000);
	setTimeout(function() {
		resolve();
	}, 1000);
	firebase.database().ref('cron-jobs/' + data.cronTaskId + '/data/').child('hourlyCount').set(data.hourlyCount + 1).then(function(){
		console.log('incremented task hourly count' + data.hourlyCount);
	});
    });

This causes the cron-task to create duplicate tasks. Again....this was all stuff that has been working flawlessly up until recently. I'm unsure if writing to the cron-task is affecting the 'lastRun' and 'nextRun'

**Comment by doowb** [comment]
Since `cron-jobs` is the endpoint where the job metadata is stored, then there is a listener on that and when it updates it will "reset" the timer that checks to see if any jobs are ready to run.

However, since at job was updated already, it should have the `nextRun` set and your change shouldn't affect anything.

One thing to point out, since your `setTimeout` calls are both `1000`, they'll run at the same time.

**Comment by matthieunelmes** [comment]
I hadn't considered that. I will retest.

**Comment by matthieunelmes** [comment]
Ok, upon further investigation I can confirm that it's the write within the queue worker, specifically referencing the cron-jobs endpoint that is causing firebase-cron to add duplicate tasks.

I am able to work around this by moving the incremental hourlyCount on a separate node, however it would be nice to get to the bottom of this.

**Comment by matthieunelmes** [comment]
Just to further expand on why I have the incremental value, it's so that I can track how many times a task has been run and then remove the task after 72 hours.

I know Firebase has transaction methods that may be more appropriate for this scenario to increment the count, however. This still relies on referencing the cron-jobs endpoint.

**Comment by doowb** [comment]
Thanks for the update. I'm going to leave this issue open because your use case is something that I've thought of before (specifically keeping track of the number of times a job has run and other metadata about a job).

A recommendation that I have about using a library like this, it's probably better to keep the data outside of the `cron-jobs` endpoint because that's being used internally by `firebase-cron` and there's no guarantee that the data structure won't change in the future (especially if I implement your suggestion above).

Would it be helpful if the job kept track of the number of times it ran and when each job was run? I'm not able to track the runtime since that's handed off to `firebase-queue` right now.

I use this library for a lot of tasks and there are other features, like one-time runs, schedule a specific time instead of the cron pattern, etc... that I want to add. I might work on this some this weekend to get those implemented since I'm working on some other architecture updates.

**Comment by matthieunelmes** [comment]
> A recommendation that I have about using a library like this, it's probably better to keep the data outside of the cron-jobs endpoint because that's being used internally by firebase-cron and there's no guarantee that the data structure won't change in the future (especially if I implement your suggestion above).

Agreed, as I have different types of tasks. I think I'll setup an entirely different endpoint of job descriptors that describe things such as how many times the job should run etc etc.

> Would it be helpful if the job kept track of the number of times it ran and when each job was run? I'm not able to track the runtime since that's handed off to firebase-queue right now.

This would be really beneficial for my specific use-case. 

> I use this library for a lot of tasks and there are other features, like one-time runs, schedule a specific time instead of the cron pattern, etc... that I want to add. I might work on this some this weekend to get those implemented since I'm working on some other architecture updates.

Sounds great, happy to help out where ever possible with testing.
